keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+This document outlines the proposal for moving kubeadm outside of the
+kubernetes/kubernetes repository into the kubernetes/kubeadm
+repository.
+
+A proposal was already done in the following Google Doc and it
+received a healthy discussion:
+https://docs.google.com/document/d/1xFAGAjfPQZrcqmapTwpJR5aUA7-2ku1asVeq3dBrjnQ
+
+## Motivation
+
+The monolith of kubernetes/kubernetes (k/k) is splitting and components
+such as kubeadm should reside in designated repositories. Multiple
+discussions during 2019 happened where the SIG Architecture Code
+Organization sub project suggested that it is time for kubeadm to
+move and that the kubeadm maintainers should start investigating
+what has to be done for this to happen.
+
+Moving kubeadm will benefit both kubernetes/kubernetes and the kubeadm
+maintainers:
+- Less noise in k/k from kubeadm issues and PRs.
+- Make it easier for the kubeadm maintainers to track kubeadm PRs.
+- No need to run the full suite of k8s pre-submit e2e jobs on kubeadm PRs.
+
+### Goals
+
+- Move the kubeadm source code to the k/kubeadm repository.
+- Preserve the current kubeadm versioning, matching the k8s versioning.
+- Continue packing kubeadm as part of the "node" tarball for Windows
+and Linux.
+- Continue having kubeadm test jobs as release-informing (-blocking in the future).
+- Continue testing kubeadm using the existing set of tools - e.g. kinder.
+- Continue to release kubeadm documentation as part of the k8s.io website
+every k8s release.
+
+### Non-Goals
+
+- Decouple kubeadm from the k8s versioning. This can only create confusion
+for the users and those depending on k8s/kubeadm version skew.
+- Not releasing kubeadm artifacts for end users to consume.
+- Use Bazel for building kubeadm once in the k/kubeadm repository.
+For a smaller repository go tooling is enough and the effort required
+for supporting different build methods is not justified.
+- Refactor the kubeadm source structure to easily distinguish between
+frontend and backend. This work is out of scope for this proposal.
+- Use the "staging" pattern from k/k to gradually move kubeadm out.
+kubeadm already does not depend on k/k and this pattern was seen as redundant.
+
+## Proposal
+
+### Actions
+
+Once all implementation details are clear between kubeadm maintainers
+and stakeholder SIGs such as Release and Docs, the following actions should
+be taken in this exact order:
+1) Update the following tracking issue to collect all
+actions: https://github.com/kubernetes/kubeadm/issues/1949
+2) Copy kubeadm to the k/kubeadm repository.
+3) Implement build / hack / verify tooling for k/kubeadm.
+4) Add pre-submit e2e jobs for k/kubeadm. Ideally there should be
+two per branch - one that verifies / builds the code base
+and one that does a full e2e using "kinder".
+5) Implement tooling in the k/release repository to build kubeadm artifacts.
```

Comment:
perhaps "...to build kubeadm artifacts independent from k/k artifacts."

Reply:
Ie: today it builds kubeadm artifacts, but this building is conflated with building other artifacts.

"Independent" also might not be the right word, since the KEP is written to have the two builds dependent in terms of branch names and version strings.

Reply:
i will clarify this point.